python-package/lightgbm/engine.py

```diff
@@ -4,21 +4,42 @@
 import copy
 from operator import attrgetter
 from pathlib import Path
+from typing import Any, Callable, Dict, List, Optional, Tuple, Union
 
 import numpy as np
 
 from . import callback
 from .basic import Booster, Dataset, LightGBMError, _ConfigAliases, _InnerPredictor, _log_warning
 from .compat import SKLEARN_INSTALLED, _LGBMGroupKFold, _LGBMStratifiedKFold
 
-
-def train(params, train_set, num_boost_round=100,
-          valid_sets=None, valid_names=None,
-          fobj=None, feval=None, init_model=None,
-          feature_name='auto', categorical_feature='auto',
-          early_stopping_rounds=None, evals_result=None,
-          verbose_eval=True, learning_rates=None,
-          keep_training_booster=False, callbacks=None):
+_LGBM_CustomObjectiveFunction = Callable[
+    [Union[List, np.ndarray], Dataset],
+    Tuple[Union[List, np.ndarray], Union[List, np.ndarray]]
+]
+_LGBM_CustomMetricFunction = Callable[
+    [Union[List, np.ndarray], Dataset],
+    Tuple[str, float, bool]
+]
+
+
+def train(
+    params: Dict[str, Any],
+    train_set: Dataset,
+    num_boost_round: int = 100,
+    valid_sets: Optional[List[Dataset]] = None,
+    valid_names: Optional[List[str]] = None,
+    fobj: Optional[_LGBM_CustomObjectiveFunction] = None,
+    feval: Optional[_LGBM_CustomMetricFunction] = None,
```

Comment:
```suggestion
    feval: Optional[Union[_LGBM_CustomMetricFunction, List[_LGBM_CustomMetricFunction]]] = None,
```

Reply:
ah, right! got it

https://github.com/microsoft/LightGBM/blob/11d7608f2d755899a88bfa1c0fbc779b9e9b3505/python-package/lightgbm/engine.py#L58